File: leet_easy_python/leet_437/leet_437b.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def pathSum(self, root, sum):
        self.targ = sum
        self.dic = {}
        self.cnt = 0
        self.pSh(root, sum,0)
        return self.cnt
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        
    def pSh(self, iter, sum, len): #pSh = pathSum helper
        len +=1
        if not iter or (iter, len) in self.dic: return
        self.dic[(iter,len)]=1
        if iter.val == sum:
            self.cnt+=1
        if iter.val == self.targ and (iter, 1) not in self.dic:
            self.dic[(iter,1)]=1
            self.cnt+=1
        self.pSh(iter.left,self.targ, 0)
        self.pSh(iter.right,self.targ, 0)
        self.pSh(iter.left,sum - iter.val, len)
        self.pSh(iter.right,sum - iter.val, len)
        """
        :type iter: TreeNode
        :type sum: int
        :type len: int
        :rtype: None
        """
        
        #still exceeds time limit

File: leet_easy_java/leet_437/leet_437b.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int pathSum(TreeNode root, int sum){
        return pathSum(root, sum, 0);
    }
    public int pathSum(TreeNode root, int sum, int flag){
        if(root == null) return 0;
        if(flag == 0){
            if(root.val == sum) return pathSum(root.right, sum,0) + pathSum(root.left, sum,0) + pathSum(root.left, sum - root.val, 1) + pathSum(root.right, sum - root.val, 1) + 1;
            return pathSum(root.right, sum, 0) + pathSum(root.left, sum, 0) + pathSum(root.left, sum - root.val, 1) + pathSum(root.right, sum - root.val, 1);
        }
        else{
            if(root.val == sum) return pathSum(root.left, sum - root.val, 1) + pathSum(root.right, sum - root.val, 1) + 1;
            return pathSum(root.left, sum - root.val, 1) + pathSum(root.right, sum - root.val, 1);
        }
    }
}